File: LeetCodePython/py46.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        tmp = []
        n = len(nums)
        result = []
        used = {}
        def helper(count):
            if count == n:
                result.append(tmp[:])
            else:
                for i in range(n):
                    if i not in used:
                        tmp.append(nums[i])
                        used[i] = 1
                        helper(count+1)
                        tmp.pop()
                        del used[i]
        helper(0)
        return result

File: LeetCodeCPlusPlus/py46.cpp
// 没有重复数字的全排列
#include <vector>
using namespace std;
class Solution {
private:
    void swap(int&a, int&b) {
        int tmp = a;
        a = b;
        b = tmp;
    }
    vector<int> tmp;
    void dfs(vector<vector<int>>&result, vector<int>&nums, int index) {
        if (index == nums.size()) {
            result.push_back(nums);
            return;
        }
        for (int i = index; i < nums.size(); i++) {
            swap(nums[i], nums[index]);
            dfs(result, nums, index+1);
            swap(nums[i], nums[index]);
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        dfs(result, nums, 0);
        return result;
    }
};